lib/rspec/rails/matchers/have_reported_error.rb

```diff
@@ -0,0 +1,271 @@
+module RSpec
+  module Rails
+    module Matchers
+      # @api private
+      # Sentinel value to distinguish between no argument passed vs explicitly passed nil.
+      # This follows the same pattern as RSpec's raise_error matcher.
+      UndefinedValue = Object.new.freeze
+
+      # @api private
+      class ErrorSubscriber
+        attr_reader :events
+
+        ErrorEvent = Struct.new(:error, :attributes)
+
+        def initialize
+          @events = []
+        end
+
+        def report(error, **attrs)
+          @events << ErrorEvent.new(error, attrs.with_indifferent_access)
+        end
+      end
+
+      # Matcher class for `have_reported_error`. Should not be instantiated directly.
+      #
+      # Provides a way to test that an error was reported to Rails.error.
+      #
+      # @api private
+      # @see RSpec::Rails::Matchers#have_reported_error
+      class HaveReportedError < RSpec::Rails::Matchers::BaseMatcher
+        # Uses UndefinedValue as default to distinguish between no argument
+        # passed vs explicitly passed nil.
+        #
+        # @param expected_error_or_message [Class, String, Regexp, nil]
+        #   Error class, message string, or message pattern
+        # @param expected_message [String, Regexp, nil]
+        #   Expected message when first param is a class
+        def initialize(expected_error_or_message = UndefinedValue, expected_message = nil)
+          @attributes = {}
+          @warn_about_nil_error = expected_error_or_message.nil?
+
+          case expected_error_or_message
+          when UndefinedValue
+            @expected_error = nil
+            @expected_message = expected_message
+          when String, Regexp
+            @expected_error = nil
+            @expected_message = expected_error_or_message
+          else
+            @expected_error = expected_error_or_message
+            @expected_message = expected_message
+          end
+        end
+
+        def with_context(expected_attributes)
+          @attributes.merge!(expected_attributes)
```

Comment:
Do we need to merge? Maybe just assign?

Reply:
Agreed, chaining another `with_context(...)` on an existing `have_reported_error.with_context(user_id: 123)`, should not overwrite the first one, at least without warning that it does.
@skatkov can you please add a spec example to solidify this behaviour?

Reply:
Currently, `.merge!` will overwrite previously defined `with_context(..)`. So I modified this code to detect possible conflicting keys and raise a warning.